### DISCUSSION TITLE
Launch on raspberry pi 4?

### DISCUSSION THREAD

**Discussion by oosavu** [bevyengine/bevy#3821] [Q&A]
Bevy version 0.6
Raspberry PI 4 (Raspberry PI OS)

Hi! Is it possible to launch Bevy on Raspberry PI 4? Has anyone done that?
It is a fresh Linux instatlation. packages installed: `alsa mesa-vulkan-drivers vulkan-mesa pkg-config libx11-dev libasound2-dev libudev-dev vulkan vulkan-headers`.

Then i run examples with `--features bevy_internal/webgl` - it works, but FPS is very slow (around 1-2)
Then i run without this flag, i get following error:
```
pi@raspberrypi:~/bevy $ cargo run --example animate_shader
    Finished dev [unoptimized + debuginfo] target(s) in 0.49s
     Running `target/debug/examples/animate_shader`
2022-01-31T00:24:34.688948Z  INFO winit::platform_impl::platform::x11::window: Guessed window scale factor: 1    
WARNING: v3dv is neither a complete nor a conformant Vulkan implementation. Testing use only.
2022-01-31T00:24:34.916315Z  INFO bevy_render::renderer: AdapterInfo { name: "V3D 4.2", vendor: 5348, device: 42, device_type: IntegratedGpu, backend: Vulkan }
2022-01-31T00:24:34.917850Z ERROR wgpu_hal::vulkan::instance: GENERAL [Loader Message (0x0)]
	terminator_CreateDevice: Failed in ICD /usr/lib/arm-linux-gnueabihf/libvulkan_broadcom.so vkCreateDevicecall    
2022-01-31T00:24:34.917977Z ERROR wgpu_hal::vulkan::instance: 	objects: (type: INSTANCE, hndl: 0x38af288, name: ?)    
2022-01-31T00:24:34.918059Z ERROR wgpu_hal::vulkan::instance: GENERAL [Loader Message (0x0)]
	vkCreateDevice:  Failed to create device chain.    
2022-01-31T00:24:34.918149Z ERROR wgpu_hal::vulkan::instance: 	objects: (type: INSTANCE, hndl: 0x38af288, name: ?)    
2022-01-31T00:24:34.918441Z ERROR wgpu::backend::direct: Error in Adapter::request_device: connection to device was lost during initialization    
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: RequestDeviceError', crates/bevy_render/src/renderer/mod.rs:110:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

What is the right way to launch bevy 0.6 on raspberry? Thanks!

**Comment by bjorn3** [comment]
> WARNING: v3dv is neither a complete nor a conformant Vulkan implementation. Testing use only.

This may be the reason why vulkan doesn't work.

**Comment by heavyrain266** [comment]
Vulkan driver for RPis `v3dv` is incomplete and many stuff (mostly shaders) will fail, It's not really possible to fully run bevy on RPis yet.

Edit: You may want to try closed source broadcom driver but not sure if it will change anything.

**Comment by RustyPyGuy** [comment]
I have been attempting to launch examples of Bevy latest on a RaspberryPi 4.  I think the issue is somewhere between bevy and wgpu requesting or expecting features that Mesa doesn't support with the v3d driver when requesting the device.  

Here's what I've done so far:

- Configured RaspberryPi OS testing with X11.  Compiled and installed the latest development Mesa drivers. Having said that, I see no problem with your Raspberry OS build.
- Verified the correct driver is running with vulkaninfo and set the `VK_ICD_FILENAMES` environment variable to match my compiled Mesa drivers.  (again, the Mesa drivers distributed with RaspberryPi OS probably work as vulkaninfo gives output)
- **Successfully ran most SaschaWillems/Vulkan examples** and verified my compiled version of Mesa is running (only the Vulkan features not implemented by the v3d chipset or driver fail to run).  This indicates that the Mesa Vulkan drivers are working.
- **Successfully ran a majority of wgpu examples** on the RaspberryPi (only the Vulkan features not implemented by the v3d chipset fail to run).  This indicates the issue is not with this layer of the graphics stack.
- **Successfully ran most winit examples.**  This indicates the issue isn't with windowing in X11.
- **All Bevy graphical examples fail** with the text `ERROR wgpu::backend::direct: Error in Adapter::request_device: connection to device was lost during initialization`
- Grepped the exact text `connection to device was lost during initialization` which is in `wgpu-core/instance.rs`  This text does not appear in Bevy.  I tried grepping some of the other unique phrases in the error, but can’t find their source.
- The only available` ash` example `cargo run --example winit `succeeds. (some searches had me suspecting ash, but I don’t think it’s the cause).  Ash also contains a separate directory of examples with separate Cargo.toml.  That requires an SDK to execute which is not available for armv7.

I also want to solve this as well for my project.  It seems to me that one course of action would be to open an issue with either Bevy or wgpu, but I’ve taken this about far as I can go.  I am hopeful that this can be fixed because the wgpu examples are largely successful.  Thoughts?

My full output:
```
2022-02-13T20:06:52.077409Z  INFO winit::platform_impl::platform::x11::window: Guessed window scale factor: 1    
2022-02-13T20:06:52.771331Z  INFO bevy_render::renderer: AdapterInfo { name: "V3D 4.2", vendor: 5348, device: 3192414163, device_type: IntegratedGpu, backend: Vulkan }
2022-02-13T20:06:52.776593Z ERROR wgpu_hal::vulkan::instance: GENERAL [Loader Message (0x0)]
        terminator_CreateDevice: Failed in ICD /usr/local/lib/libvulkan_broadcom.so vkCreateDevicecall    
2022-02-13T20:06:52.776726Z ERROR wgpu_hal::vulkan::instance:   objects: (type: INSTANCE, hndl: 0x3d0b570, name: ?)    
2022-02-13T20:06:52.776807Z ERROR wgpu_hal::vulkan::instance: GENERAL [Loader Message (0x0)]
        vkCreateDevice:  Failed to create device chain.    
2022-02-13T20:06:52.776895Z ERROR wgpu_hal::vulkan::instance:   objects: (type: INSTANCE, hndl: 0x3d0b570, name: ?)    
2022-02-13T20:06:52.777188Z ERROR wgpu::backend::direct: Error in Adapter::request_device: connection to device was lost during initialization    
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: RequestDeviceError', crates/bevy_render/src/renderer/mod.rs:215:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

```

**Reply by cart** [reply]
If you try setting  `WgpuSettingsPriority` to `WebGL2` on the `WgpuSettings` resource, does that help? (note that it won't actually use webgl2, just downlevel some feature requests).

**Reply by RustyPyGuy** [reply]
Yes! That does it.  Bevy works on RaspberryPi 4! Thank you cart!

The easy solution is to set the environment variable:
`WGPU_SETTINGS_PRIO="webgl2"` 
or depending on commit version including Bevy 0.6:
`WGPU_OPTIONS_PRIO="webgl2"`

It uses Vulkan as expected.  This inspires me to contribute to the unwrap() -> expect() issue to give some useful info about this when it panics.  I'm not sure I want to take on the entire render crate myself, but perhaps!

@oosavu Give it a try!

**Comment by nickelser** [comment]
Was anyone able to get a graphical Bevy project running with a reasonable framerate on a Pi 4?

The examples (even run with `--release`) are basically unplayable, and simple 2d games run at < 10 fps even at very low resolution. It does seem to be CPU-bound. Maybe I'm missing a driver setting or other configuration...?

**Comment by BorisBoutillier** [comment]
For reference for people coming back to this thread.
I have been able to run bevy projects on the Raspberry Pi  4B:
- I used a default RaspianOS 64 bit image with `rpi-imager`, which as of today uses Debain 12 (bookworm).
- I installed  the now fully supported vulkan drivers through a simple  `sudo apt install mesa-vulkan-drivers`
- I cross compiled to `aarch64-uknown-linux-gnu` using the [cross](https://github.com/cross-rs/cross) project with a custom Dockerfile:
```
 FROM ghcr.io/cross-rs/aarch64-unknown-linux-gnu:latest

RUN dpkg --add-architecture arm64 && \
    apt-get update && \
    apt-get install --assume-yes libasound2-dev:arm64 libx11-dev:arm64 libudev-dev:arm64 

ENV PKG_CONFIG_LIBDIR_aarch64_unknown_linux_gnu=/usr/lib/aarch64-linux-gnu/pkgconfig
```

I have been able to play the _breakout_ example game at around 20 fps, but something that look like dropped  frames  every second ( did not investigate). Same for the _bevymark_  example, significantly dropping from the starting 40 FPS after less than 2000 spawns.  
3D scenes also work, I did not check FPS, but probably low.

So don't expect to heavily play on a Raspberry 4B :)
